hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/grantapproval/GrantApprovalDecoder.java

```diff
@@ -0,0 +1,129 @@
+import com.hedera.node.app.service.contract.impl.exec.systemcontracts.hts.AddressIdConverter;
+import com.hedera.node.app.service.contract.impl.exec.systemcontracts.hts.HtsCallAttempt;
+import com.hedera.node.app.service.contract.impl.utils.ConversionUtils;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.math.BigInteger;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+@Singleton
+public class GrantApprovalDecoder {
+
+    @Inject
+    public GrantApprovalDecoder() {
+        // Dagger2
+    }
+
+    public TransactionBody decodeGrantApproval(@NonNull final HtsCallAttempt attempt) {
+        final var call = GrantApprovalTranslator.GRANT_APPROVAL.decodeCall(attempt.inputBytes());
+        return TransactionBody.newBuilder()
+                .transactionID(
+                        TransactionID.newBuilder().accountID(attempt.senderId()).build())
+                .cryptoApproveAllowance(
+                        grantApproval(attempt.addressIdConverter(), call.get(0), call.get(1), call.get(2)))
+                .build();
+    }
+
+    public TransactionBody decodeGrantApprovalNFT(@NonNull final HtsCallAttempt attempt) {
+        final var call = GrantApprovalTranslator.GRANT_APPROVAL_NFT.decodeCall(attempt.inputBytes());
+        return TransactionBody.newBuilder()
+                .transactionID(
+                        TransactionID.newBuilder().accountID(attempt.senderId()).build())
+                .cryptoApproveAllowance(
+                        grantApprovalNFT(attempt.addressIdConverter(), call.get(0), call.get(1), call.get(2)))
+                .build();
+    }
+
+    public TransactionBody decodeErcGrantApproval(@NonNull final HtsCallAttempt attempt) {
+        final var call = GrantApprovalTranslator.ERC_GRANT_APPROVAL.decodeCall(attempt.inputBytes());
+        return TransactionBody.newBuilder()
+                .transactionID(
+                        TransactionID.newBuilder().accountID(attempt.senderId()).build())
+                .cryptoApproveAllowance(ERCGrantApproval(
+                        attempt.addressIdConverter(),
+                        requireNonNull(attempt.redirectTokenId()),
+                        call.get(0),
+                        call.get(1),
+                        requireNonNull(attempt.redirectTokenType())))
+                .build();
+    }
+
+    private CryptoApproveAllowanceTransactionBody grantApproval(
+            @NonNull final AddressIdConverter addressIdConverter,
+            @NonNull final Address token,
+            @NonNull final Address spender,
+            @NonNull final BigInteger amount) {
+        return CryptoApproveAllowanceTransactionBody.newBuilder()
+                .tokenAllowances(TokenAllowance.newBuilder()
+                        .tokenId(ConversionUtils.asTokenId(token))
+                        .spender(addressIdConverter.convert(spender))
+                        .amount(amount.longValue())
```

Comment:
Do you know if the amount is ever validated to being not negative for fungible tokens?  I presume it may be done in the token service but it might help to check.

Reply:
thanks for checking.